### PR TITLE
Add Index-64 API as extended API with _64 suffix for LAPACKE

### PR DESCRIPTION
**Pull request by mkrainiuk**
**Description**
This PR enables extended 64-bit Integer API with `_64` suffix for LAPACKE and its dependencies (https://github.com/Reference-LAPACK/lapack/issues/666#).

As it was suggested in https://github.com/Reference-LAPACK/lapack/issues/666# Fortran API renaming happens via common header file `SRC/lapack_64.h` instead of compilation flag. So potentially same file could be reused in other projects. But this approach requires Fortran compiler to support preprocessing. So I had to remove new header file from Fortran files (the header is added during the build to the local copy of each file 32d3f56cd76be2378af3558c54b0e8a15772c8dd) and enabled new extended API only for GNU and Intel compilers (fac7a322dbb1f981927b51b78e3ff7933688dd83), based on my experiments LLVM compiler has some problems with preprocessing Fortran.

Please note, for the migration simplification (and it was also one of Julia language community feedback) the extended API with `_64` suffix is added to ALL functionality, including functions that don't have Integer type in the API.

Most of the changes in LAPACK address the problem with FORTRAN 77 line length limitation. All FORTRAN 77 lines with LAPACK API and length > 69 symbols failed because of 3 more symbols with `_64`. Since I'm not sure this project uses any linter tools, I run my local script for simply moving the last argument from the affected line to the new line, all changes made by the script are grouped to:
* 185aa3deec9852d30768beb7485bec33186ea345 for LAPACK
* 3970a1afaa9719f7bb17cf6e78a10a18e436ff27 for MATGEN
* and some manual changes after rebasing on the latest master 8da80f0b32001a4353a881ca452609abd27b2c35

Let me know if the code should be updated with some linter tool.

There was also a problem with missed MOD files in LAPACK parallel build (the file that needs the MOD file could be compiled before the file that has the required modules). PR includes the fix for it b658fd6897fe52bd86726adf845f8d55d4129c56.

**Checklist**

- [x] The documentation has been updated. LAPACK documentation is in PDF format, I can't update.
- [x] If the PR solves a specific issue, it is set to be closed on merge. It addresses LAPACK part in https://github.com/Reference-LAPACK/lapack/issues/666#
- [x] Local testing and new LAPACKE _64 examples passed

`-fdefault-integer-8` and `-Werror=conversion` caused build errors in CI because of implicit casting INTEGER to REAL or COMPLEX in many files. Existing flag "-DBUILD_INDEX64=ON" has the same problem with this flag. The following commits address this problem by changing casting from implicit to explicit:
* In MATGEN adcd82df8845d91c6c1ee60c3ab08950f8564681
* In LAPACK f27db1d2cd19199b84e1cadf9e4f438100aa51e1

All CI steps passed after rebasing, `continuous-integration/appveyor/pr` step problem was cause by LLVM Fortran compiler, since now the extended API is disabled for this compiler, this step also passed.

### CODE REVIEW

**Comment by weslleyspereira** [comment]
Hi @mkrainiuk. Would you be able to merge to/rebase with the current master? The script for the Github Actions needed to be updated. Thanks

**Comment by mkrainiuk** [comment]
> Hi @mkrainiuk. Would you be able to merge to/rebase with the current master? The script for the Github Actions needed to be updated. Thanks

Thanks @weslleyspereira! After rebasing everything is passing except for `continuous-integration/appveyor/pr` I'm investigating why.

**Review by angsch — COMMENTED** [review]
_(I have only looked at the examples - this is not a proper review)_

 For the `DGESV` examples, wouldn't it be good to just hard-code a sample matrix? I find the number of lines dedicated to command line argument parsing, and memory handling a bit too high for a what I assume shall be a simple example of how to call the LAPACKE function. Further, for the sake of readability, please fix the indentation and line breaks of the `DGESV` examples.

File: LAPACKE/example/example_DGESV_colmajor_64.c [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+		    }
+			if( strcmp( argv[i], "-nrhs" ) == 0 ) {
+				nrhs  = atoi(argv[i+1]);
+				i++;
+			}
+		}
+
+        /* Initialization */
+        lda=n, ldb=n;
+		A = (double *)malloc(n*n*sizeof(double)) ;
```
[the remark this answers is not in the document]
Comment (mkrainiuk) [reply LAPACKE/example/example_DGESV_colmajor_64.c]: Thanks for catching it, fixed both original examples and the ones with extended API.

**Comment by mkrainiuk** [comment]
@langou since all tests passed now please let me know if there is something else I need to complete in order to merge the PR?

**Review by langou — APPROVED** [review]

**Comment by langou** [comment]
Thanks for the reminder @mkrainiuk

**Pull request merged**